src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs

```diff
@@ -206,53 +254,214 @@ public static FormulaValue Text(EvalVisitor runner, EvalVisitorContext context,
+                          .Replace("\u0006", hasAmPm ? dateTime.ToString("%h", culture) : dateTime.ToString("%H", culture));
+
+            // Minute component
+            format = _minutesDetokenizeRegex.Replace(format, dateTime.ToString("mm", culture))
+                          .Replace("\u000A", dateTime.ToString("%m", culture));
+
+            // Second component
+            format = _secondsDetokenizeRegex.Replace(format, dateTime.ToString("ss", culture))
+                          .Replace("\u0008", dateTime.ToString("%s", culture));
+
+            // Milliseconds component
+            format = _milisecondsDetokenizeRegex.Replace(format, dateTime.ToString("fff", culture))
+                          .Replace("\u000E\u000E", dateTime.ToString("ff", culture))
+                          .Replace("\u000E", dateTime.ToString("%f", culture));
+
+            // AM/PM component
+            format = format.Replace("\u0001", dateTime.ToString("tt", culture))
+                           .Replace("\u0002", dateTime.ToString("%t", culture).ToLower());
+
+            return format;
+        }
+
+        private static string TokenizeDatetimeFormat(string format, CancellationToken cancellationToken)
         {
-            var pattern = @"^(?<openAMPM>\s*t+\s*)? " +
-                             @"(?(openAMPM) h+(?<nonHours>[^ht]+)$ " +
-                             @"| \s*h+(?<nonHours>[^ht]+)\s*t+)";
-            return Regex.Replace(format, pattern, "HH${nonHours}", RegexOptions.IgnorePatternWhitespace);
+            // Temporary replacements to avoid collisions with upcoming month names, etc.
+            format = _ampmReplaceRegex.Replace(format, "\u0001");
+            format = _apReplaceRegex.Replace(format, "\u0002");
+
+            // Find all "m" chars for minutes, before seconds
+            var match = _minutesBeforeSecondsRegex.Match(format);
+            while (match.Success)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                format = format.Substring(0, match.Index) + "\u000A" + format.Substring(match.Index + 1);
+                match = _minutesBeforeSecondsRegex.Match(format);
+            }
+
+            // Find all "m" chars for minutes, after hours
+            match = _minutesAfterHoursRegex.Match(format);
+            while (match.Success)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var afterHourFormat = format.Substring(match.Index);
+                var minuteAfterHourPosition = _minutesRegex.Match(afterHourFormat);
+                var pos = match.Index + minuteAfterHourPosition.Index;
+
+                format = format.Substring(0, pos) + "\u000A" + format.Substring(pos + 1);
+
+                match = _minutesAfterHoursRegex.Match(format);
+            }
+
+            var sb = new StringBuilder();
+            foreach (var c in format)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
```

Comment:
>cancellationToken.ThrowIfCancellationRequested(); [](http://example.com/codeflow?start=16&length=49)

I don't think we need this code here. This is a very fast loop - it goes through at most 100 characters and appends a character into a StringBuilder instance. No need to check for cancellation for every single char.